client/state/data-layer/wpcom/jetpack-onboarding/index.js

```diff
@@ -116,13 +116,46 @@ export const saveJetpackOnboardingSettings = ( { dispatch, getState }, action )
 export const handleSaveSuccess = ( { dispatch }, { siteId, settings } ) =>
 	dispatch( saveJetpackOnboardingSettingsSuccess( siteId, settings ) );
 
-export const announceSaveFailure = ( { dispatch }, { siteId } ) =>
+export const announceSaveFailure = ( { dispatch }, { siteId } ) => {
 	dispatch(
 		errorNotice( translate( 'An unexpected error occurred. Please try again later.' ), {
 			id: `jpo-notice-error-${ siteId }`,
 			duration: 5000,
 		} )
 	);
+};
+
+export const retryOrAnnounceSaveFailure = ( { dispatch }, action ) => {
+	const MAX_WOOCOMMERCE_INSTALL_RETRIES = 3;
+	const { settings, siteId, type, meta: { dataLayer } } = action;
+	const { error, retryCount = 0 } = dataLayer;
+
+	// If we got a timeout on WooCommerce installation, try again (up to 3 times),
+	// since it might just be a slow server that actually ends up installing it
+	// properly, in which case a subsequent request will return 'success'.
+	if (
+		settings.installWooCommerce !== true ||
+		error.error !== 'http_request_failed' ||
+		retryCount > MAX_WOOCOMMERCE_INSTALL_RETRIES
+	) {
+		return announceSaveFailure( { dispatch }, { siteId } );
+	}
+
+	// We cannot use `extendAction( action, ... )` here, since `meta.datayLayer` now includes error information,
+	// which we would propagate, causing the data layer to think there's been an error on the subsequent attempt.
+	// Instead, we have to re-assemble our action.
```

Comment:
Should there be a generic tool that strips error information from `meta.dataLayer` for cases like this?